### PR TITLE
Backoff after querying all the servers on fast failover

### PR DESCRIPTION
**Pull request by fsamuel-bs**
**Goals (and why)**: Implement the fix proposed on #1799.

**Implementation Description (bullets)**: When an exception triggers fast failover, check if we’ve queried all the nodes in the cluster. If so, sleep for 1s and then re-query the nodes.

**Concerns (what feedback would you like?)**: I tried making the minimal change. Did I change any behavior that should not be changed?

**Where should we start reviewing?**: FailoverFeignTarget.

**Priority (whenever / two weeks / yesterday)**: Whenever, but soon, to fix the test flakiness.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1939)

### CODE REVIEW

File: atlasdb-config/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java [hunk 1]
```diff
@@ -143,11 +147,16 @@ private void checkAndHandleFailure(RetryableException ex) {
...
 
 
-    private void pauseForBackOff(RetryableException ex) {
-        double pow = Math.pow(
+    private void pauseForBackoff(RetryableException ex) {
+        double pauseTime = Math.pow(
                 GOLDEN_RATIO,
                 numSwitches.get() * failuresBeforeSwitching + failuresSinceLastSwitch.get());
-        long timeout = Math.min(maxBackoffMillis, Math.round(pow));
+        pauseForBackoff(ex, Math.round(pauseTime));
```
Comment (schlosna) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java]: may want to incorporate some jitter in the backoff, see https://www.awsarchitectureblog.com/2015/03/backoff.html

**Comment by gsheasby** [comment]
I've had a look at test logs from before and after this change. Before, we see failed fresh-timestamp requests many (at least 16) times per second:
```
3406/ete1.log:ete1_1       | 172.19.0.3 - - [15/May/2017:12:37:03 +0000] "POST /timestamp/fresh-timestamp HTTP/1.1" 503 7503 "-" "unknown-atlasdb (unknown)" 49
3406/ete1.log:ete1_1       | 172.19.0.3 - - [15/May/2017:12:37:03 +0000] "POST /timestamp/fresh-timestamp HTTP/1.1" 503 7503 "-" "unknown-atlasdb (unknown)" 9
3406/ete1.log:ete1_1       | 172.19.0.3 - - [15/May/2017:12:37:03 +0000] "POST /timestamp/fresh-timestamp HTTP/1.1" 503 7503 "-" "unknown-atlasdb (unknown)" 3
3406/ete1.log:ete1_1       | 172.19.0.3 - - [15/May/2017:12:37:03 +0000] "POST /timestamp/fresh-timestamp HTTP/1.1" 503 7503 "-" "unknown-atlasdb (unknown)" 3
3406/ete1.log:ete1_1       | 172.19.0.3 - - [15/May/2017:12:37:03 +0000] "POST /timestamp/fresh-timestamp HTTP/1.1" 503 7503 "-" "unknown-atlasdb (unknown)" 4
3406/ete1.log:ete1_1       | 172.19.0.3 - - [15/May/2017:12:37:03 +0000] "POST /timestamp/fresh-timestamp HTTP/1.1" 503 7503 "-" "unknown-atlasdb (unknown)" 3
```

After, these requests are spread out, so there is only one per second:
```
3532/ete1.log:ete1_1       | 172.19.0.3 - - [22/May/2017:18:37:15 +0000] "POST /timestamp/fresh-timestamp HTTP/1.1" 503 7503 "-" "unknown-atlasdb (unknown)" 40
3532/ete1.log:ete1_1       | 172.19.0.3 - - [22/May/2017:18:37:16 +0000] "POST /timestamp/fresh-timestamp HTTP/1.1" 503 7553 "-" "unknown-atlasdb (unknown)" 6
3532/ete1.log:ete1_1       | 172.19.0.4 - - [22/May/2017:18:37:17 +0000] "POST /timestamp/fresh-timestamp HTTP/1.1" 503 7553 "-" "unknown-atlasdb (unknown)" 6
3532/ete1.log:ete1_1       | 172.19.0.5 - - [22/May/2017:18:37:18 +0000] "POST /timestamp/fresh-timestamp HTTP/1.1" 503 7553 "-" "unknown-atlasdb (unknown)" 6
```

File: atlasdb-config/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java [hunk 2]
```diff
@@ -97,7 +97,11 @@ public void continueOrPropagate(RetryableException ex) {
 
         checkAndHandleFailure(ex);
         if (retryBehaviour.shouldBackoffAndTryOtherNodes()) {
-            pauseForBackOff(ex);
+            if (failoverCount.get() % servers.size() == 0) {
+                pauseForBackoff(ex, 1000L);
```
Comment (gsheasby) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java]: We should make this backoff configurable. 1 second might be too conservative for a default. We could figure out how long, on average, a leadership election takes, and then wait for that amount of time.

File: docs/source/release_notes/release-notes.rst [hunk 3]
```diff
@@ -55,6 +55,13 @@ develop
            There is no issue with having tables with different values for ``gc_grace_seconds``, and this can be updated at any time.
            (`Pull Request <https://github.com/palantir/atlasdb/pull/1726>`__)
 
+    *    - |fixed|
+         - A 1s timeout has been added to the client when the client has queried all the servers of a cluster and received a NotCurrentLeaderException.
```
Comment (gsheasby) [inline docs/source/release_notes/release-notes.rst]: This release note is a bit wordy. It's probably enough to keep the first sentence and add a second: "This should reduce contention during leadership elections".

If we make the timeout configurable, we should add a link to the docs here.

Also, we should link to this PR.

File: atlasdb-config/src/test/java/com/palantir/atlasdb/http/FailoverFeignTargetTest.java [hunk 4]
```diff
@@ -146,6 +151,26 @@ public void blockingTimeoutExceptionResetsFailureCount() {
...
+            spiedTarget.url();
+            spiedTarget.continueOrPropagate(EXCEPTION_WITH_RETRY_AFTER);
+        }
+
+        ArgumentCaptor<Long> argument = ArgumentCaptor.forClass(Long.class);
+        verify(spiedTarget, times(CLUSTER_SIZE)).pauseForBackoff(any(), argument.capture());
+
+        List<Long> arguments = argument.getAllValues();
+        assertThat(arguments.size()).isEqualTo(CLUSTER_SIZE);
+        for (int i = 0; i < CLUSTER_SIZE - 1; i++) {
```
Comment (gsheasby) [inline atlasdb-config/src/test/java/com/palantir/atlasdb/http/FailoverFeignTargetTest.java]: If we specify a cluster size for this test of 3 (i.e. without using the CLUSTER_SIZE constant), we can write a single assert to verify that `arguments.containsExactly(1L, 1L, 1000L)` (pseudo-test-code, actual method required may be different).

**Comment by gsheasby** [comment]
@gmaretic, could you check my working, please?

File: atlasdb-config/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java (hunk 1 above)
Comment (gmaretic) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java]: This is pretty awesome, and we should implement it, but can also be part of a future PR if we want to get this in ASAP

File: atlasdb-config/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java [hunk 5]
```diff
@@ -97,7 +98,11 @@ public void continueOrPropagate(RetryableException ex) {
 
         checkAndHandleFailure(ex);
         if (retryBehaviour.shouldBackoffAndTryOtherNodes()) {
-            pauseForBackOff(ex);
+            if (failoverCount.get() % servers.size() == 0) {
```
Comment (gmaretic) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java]: Should this also happen if `failoverCount.get() == 0`?

Comment (gsheasby) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java]: No.
`failoverCount.get()` should be greater than 0, but I've added a check anyway.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java (hunk 1 above)
Comment (gsheasby) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java]: Filed the improvement as part of #1940.
